### PR TITLE
#140 & #220 show submission fail

### LINKED ISSUE #140: "Submit" application should wait for Trigger on application to be Null before confirming successful Submission
Otherwise show "Problem" message.

### LINKED ISSUE #220: Create a more general modal alert system that can be used all throughout the application - replace use in ApplicationPageWrapper
From [PR#194](#215) description:
Run revalidateAll when clicking "Submit" and only submit form if it passes. If not, it re-directs to the first invalid response as above. TO-DO: we need some better notification here so the user knows why they're suddenly whisked away when they thought they were submitting, but I thought it best to leave until we've implemented the more general modal alert system.

### PR DESCRIPTION
**Pull request by nmadruga**
Fixes #140 & fixes #220

~~Branched off PR #234, please merge that one before this one!~~ (Done)

~~To be used with back-end running PR [138F](https://github.com/openmsupply/application-manager-server/pull/163)~~ (Merged)

## Changes
- The awaiting part was already done as part of PR #234 
- Separate existing modal window on `ApplicationPageWrapper` in component to be reused by `ApplicationOverview`
- Show warning message in case there is a problem found in any response during the **submission step**.

## Existing problems - not resolved
- ~~There is a problem with checking for unique user names - it considers the username entered in the current application and throws an error when trying to submit on the first time (which is actually good for testing the Modal) - Need a separated issue.~~ **SOLVED**
- Sometimes the warning is showing right after clicking to create the new application!

### CODE REVIEW

**Review by CarlosNZ — CHANGES_REQUESTED** [review]
> ~~There is a problem with checking for unique user names - it considers the username entered in the current application and throws an error when trying to submit on the first time (which is actually good for testing the Modal) - Need a separated issue.~~ SOLVED

This is not quite solved. (Note: I'm using it with latest master merged back in, with my evaluator changes.) Sometimes it works, sometimes it doesn't. Can't quite figure out what the conditions are.

**Comment by CarlosNZ** [comment]
> > ~There is a problem with checking for unique user names - it considers the username entered in the current application and throws an error when trying to submit on the first time (which is actually good for testing the Modal) - Need a separated issue.~ SOLVED
> 
> This is not quite solved. (Note: I'm using it with latest master merged back in, with my evaluator changes.) Sometimes it works, sometimes it doesn't. Can't quite figure out what the conditions are.

I _think_ this part is solved for now. We should keep an eye on it.

However, there's a problem with the reverse situation, in that the application is submitted fine (no modal) even when it should have failed submission validation.

Steps to reproduce:
- Fill in User Registration form
- Get to Summary page
- Manually edit the database to change an existing username to the new one you're trying to submit
- Submit form -- it goes through correctly when it shouldn't have.
- If you go back to the application page, it correctly re-routes and shows the error, so it's doing the revalidation correctly then.

**Comment by CarlosNZ** [comment]
> However, there's a problem with the reverse situation, in that the application is submitted fine (no modal) even when it should have failed submission validation.

This is fixed now. Problem was that `setInvalidStep` in `revalidateAndUpdate` is async, so this code in `handleSubmit` was causing problems:
```
    await revalidateAndUpdate()
    if (!invalidStep) {
      await submit()
      push(`/application/${serialNumber}/submission`)
```
The `invalidStep` value wasn't been set quick enough, so it was still `undefined` for `if(!invalidStep)`. I've changed it so `revalidateAndUpdate` returns `allValid`, and then done this:
```
    const allValid = await revalidateAndUpdate()
    if (allValid) {
      await submit()
      push(`/application/${serialNumber}/submission`)
    }
```

**Comment by nmadruga** [comment]
Thank you for fixing this, Carl!

I'll reply to your suggestions on this PR.

File: src/components/Main/ModalWarning.tsx [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import React, { useState } from 'react'
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply src/components/Main/ModalWarning.tsx]: Done

File: src/components/Main/ModalWarning.tsx [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+interface ModalWarningProps {
+  showModal: ModalProps
+  setShowModal: (props: ModalProps) => void
+  onClickAction?: Function
+}
+
+const ModalWarning: React.FC<ModalWarningProps> = ({ showModal, setShowModal, onClickAction }) => {
+  const hideModal = () => setShowModal(modalInitialValue)
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply src/components/Main/ModalWarning.tsx]: I have done a little improvement on that now. Please have a look.

I would like to use the ModalProps for setting actions for the Modal window, but that's still a TODO, I'm not entirely sure yet how to use it, but I think we could be using that to make it more generic in the future.

File: src/components/Main/ModalWarning.tsx [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        <Button
+          color="green"
+          inverted
+          onClick={() => {
+            hideModal()
+            if (onClickAction) onClickAction()
+          }}
+        >
+          <Icon name="checkmark" />
+          {strings.BUTTON_OPTION_YES}
```
[the remark this answers is not in the document]
Comment (nmadruga) [reply src/components/Main/ModalWarning.tsx]: For now I'm defining in each `message`. I don't really like this approach either, but I think we can improve this Modal component as we go.

File: src/utils/messages.ts [hunk 4]
```diff
@@ -2,6 +2,12 @@ export default {
   VALIDATION_FAIL: {
     title: 'Validation failed',
     message: 'Please fix invalid fields before continuing',
+    option: 'Yes',
```
Comment (CarlosNZ) [inline src/utils/messages.ts:5]:
```suggestion
    option: 'OK',
```
The norm for these kind of buttons is "OK", I think.

File: src/utils/messages.ts [hunk 5]
```diff
@@ -2,6 +2,12 @@ export default {
...
+  },
+  SUBMISSION_FAIL: {
+    title: 'Problem to submit application',
```
Comment (CarlosNZ) [inline src/utils/messages.ts:8]:
```suggestion
    title: 'Problem submitting application',
```

File: src/utils/messages.ts [hunk 6]
```diff
@@ -2,6 +2,12 @@ export default {
...
+    message: 'Please fix invalid fields before continuing',
+    option: 'Ok',
```
Comment (CarlosNZ) [inline src/utils/messages.ts:10]:
```suggestion
    option: 'OK',
```

**Review by CarlosNZ — APPROVED** [review]
Great, it'd be good if you had a look at the suggested changes to the messages, but nothing major.

Any existing weirdness with the Submission page (sometimes) not loading properly I'll address as part of #247.

**Pull request merged**